app/modules/net.c

```diff
@@ -10,1364 +10,757 @@
+    *refptr = LUA_NOREF;
+  } else {
+    return luaL_error(L, "invalid callback function");
   }
-
-  host_ip.addr = 0;
-  if(ESPCONN_OK == espconn_gethostbyname(pesp_conn, domain, &host_ip, net_dns_found))
-    net_dns_found(domain, &host_ip, pesp_conn);  // ip is returned in host_ip.
-
-
-  return 0;  
+  return 0;
 }
 
-// Lua: net.dns.resolve( domain, function(ip) )
-static int net_dns_static( lua_State* L )
-{
-  const char *mt = "net.socket";
-  if (!lua_isstring( L, 1 ))
-    return luaL_error( L, "wrong parameter type (domain)" );
-  
-  int rfunc = LUA_NOREF; //save reference to func
-  if (lua_type(L, 2) == LUA_TFUNCTION || lua_type(L, 2) == LUA_TLIGHTFUNCTION){
-    rfunc = luaL_ref(L, LUA_REGISTRYINDEX);
+// Lua: client:send(data, function(c)), socket:send(port, ip, data, function(s))
+int net_send( lua_State *L ) {
+  lnet_userdata *ud = net_get_udata(L);
+  if (!ud || ud->type == TYPE_TCP_SERVER)
+    return luaL_error(L, "invalid user data");
+  ip_addr_t addr;
+  uint16_t port;
+  const char *data;
+  size_t datalen = 0;
+  int stack = 2;
+  if (ud->type == TYPE_UDP_SOCKET) {
+    size_t dl = 0;
+    port = luaL_checkinteger(L, stack++);
+    if (port == 0) return luaL_error(L, "need port");
+    const char *domain = luaL_checklstring(L, stack++, &dl);
+    if (!domain) return luaL_error(L, "need IP address");
+    if (!ipaddr_aton(domain, &addr)) return luaL_error(L, "invalid IP address");
+  }
+  data = luaL_checklstring(L, stack++, &datalen);
+  if (!data || datalen == 0) return luaL_error(L, "no data to send");
+  if (lua_isfunction(L, stack) || lua_islightfunction(L, stack)) {
+    lua_pushvalue(L, stack++);
+    luaL_unref(L, LUA_REGISTRYINDEX, ud->client.cb_sent_ref);
+    ud->client.cb_sent_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  }
+  if (ud->type == TYPE_UDP_SOCKET && !ud->pcb) {
+    ud->udp_pcb = udp_new();
+    if (!ud->udp_pcb)
+      return luaL_error(L, "cannot allocate PCB");
+    udp_recv(ud->udp_pcb, net_udp_recv_cb, ud);
+    addr.addr = 0;
+    err_t err = udp_bind(ud->udp_pcb, &addr, 0);
+    if (err != ERR_OK) {
+      udp_remove(ud->udp_pcb);
+      ud->udp_pcb = NULL;
+      return luaL_lwip_checkerr(err, L);
+    }
+    if (ud->self_ref == LUA_NOREF) {
+      lua_pushvalue(L, 1);
+      ud->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+    }
+  }
+  if (!ud->pcb || ud->self_ref == LUA_NOREF)
+    return luaL_error(L, "not connected");
+  err_t err;
+  if (ud->type == TYPE_UDP_SOCKET) {
+    struct pbuf *pb = pbuf_alloc(PBUF_TRANSPORT, datalen, PBUF_RAM);
+    pbuf_take(pb, data, datalen);
+    err = udp_sendto(ud->udp_pcb, pb, &addr, port);
+    pbuf_free(pb);
+    if (ud->client.cb_sent_ref != LUA_NOREF) {
+      lua_rawgeti(L, LUA_REGISTRYINDEX, ud->client.cb_sent_ref);
+      lua_rawgeti(L, LUA_REGISTRYINDEX, ud->self_ref);
+      lua_call(L, 1, 0);
+    }
+  } else if (ud->type == TYPE_TCP_CLIENT) {
+    err = tcp_write(ud->tcp_pcb, data, datalen, TCP_WRITE_FLAG_COPY);
   }
```

Comment:
`err` seems to only be assigned, never checked. Is there a call missing to `luaL_lwip_checkerr()` before the return, or is it intentional that we're quiet about errors here?